src/IECore/StreamIndexedIO.cpp

```diff
@@ -640,18 +714,97 @@ class StreamIndexedIO::Node
 
 		// Returns the named child directory node or NULL if not existent. Loads the subindex for the child nodes (if applicable).
 		DirectoryNode* directoryChild( const IndexedIO::EntryID &name ) const;
+
 		/// returns information about the Data node
-		inline bool dataChildInfo( const IndexedIO::EntryID &name, size_t &offset, size_t &size ) const;
+		bool dataChildInfo( const IndexedIO::EntryID &name, Info &info ) const;
 
 		DirectoryNode* addChild( const IndexedIO::EntryID & childName );
-		void addDataChild( const IndexedIO::EntryID & childName, IndexedIO::DataType dataType, size_t arrayLen, size_t offset, size_t size );
+		void addDataChild(
+			const IndexedIO::EntryID &childName, IndexedIO::DataType dataType, size_t arrayLen, size_t offset, size_t size, size_t decompressedSize
+		);
 
 		void removeChild( const IndexedIO::EntryID &childName, bool throwException = true );
 
 		StreamIndexedIO::IndexPtr m_idx;
 		DirectoryNode *m_node;
 };
 
+//! Small scoped class to read from a given data block in a file, 
+//! decompressing if required.
+class StreamIndexedIO::Reader
+{
+	public:
+
+		//! If an outputBuffer is supplied then it has to be large enough to store info.decompressedSize bytes of data
+		//! and if one isn't supplied then a suitably sized buffer is created and freed on destruction.
+		Reader( StreamIndexedIO::StreamFile &f, const Node::Info &info, int threadCount = 1, char *outputBuffer = nullptr )
+			: m_data( nullptr ),
+			m_decompressedData( outputBuffer ),
+			m_size( info.size ),
+			m_decompressedSize( info.decompressedSize ),
+			m_ownDecompressedData( outputBuffer == nullptr )
+		{
+			if( m_ownDecompressedData )
+			{
+				m_decompressedData = new char[m_decompressedSize];
+			}
+
+			if( isCompressed() )
+			{
+				m_data = new char[info.size];
+				f.read( m_data, info.size, info.offset );
+
+				int bloscResult = blosc_decompress_ctx( m_data, m_decompressedData, m_decompressedSize, threadCount );
+
+				if( bloscResult <= 0 )
+				{
+					throw IECore::IOException( "StreamIndexedIO::Reader - Corrupted compressed archive" );
+				}
+			}
+			else
+			{
+				f.read( m_decompressedData, info.size, info.offset );
+			}
+		}
+
+		~Reader()
+		{
+			if ( m_data )
+			{
+				delete[] m_data;
+			}
+
+			if( m_decompressedData && m_ownDecompressedData )
+			{
+				delete[] m_decompressedData;
+			}
+		}
+
+		char *data() const
+		{
+			if( m_decompressedData )
+			{
+				return m_decompressedData;
+			}
+			else
+			{
+				return m_data;
+			}
+		}
+
+		bool isCompressed() const
+		{
+			return m_size != m_decompressedSize;
```

Comment:
Just checking, this is valid because if we tried to compress during write and it doesn't get smaller, we write uncompressed data instead, correct?